examples/containers/multi/helpers/driver.sh

```diff
@@ -0,0 +1,222 @@
+#     to receive requests, and the exporter actively exporting data
+function start_new_nodes() {
+
+    local num_nodes=$1
+    local nodes_ready=() 
+
+    echo "---- Starting $num_nodes Couchbase Server and Exporter nodes ----"
+
+    local i=0
+    for ((i; i<num_nodes; i++)); do
+        docker run -d --rm --name "node$i" --hostname="node$i.local" --network=multi_shared_network \
+        -p $((8091+i)):8091 "cbs_server_exp" > /dev/null
+        nodes_ready+=(false)
+    done
+
+    # Simple block until all nodes ready
+    echo "Waiting for nodes to come up..." && sleep "$num_nodes"
+    while true; do
+        local j=0
+        for ((j; j<num_nodes; j++)); do
+            if ! ${nodes_ready[$j]}; then
+                if docker exec "node$j" curl -fs localhost:8091 > /dev/null; then
+                    nodes_ready[$j]=true
+                    echo "Node $j ready!"
+                fi
+            fi
+        done
+
+        ready=true
+        for b in "${nodes_ready[@]}"; do if ! $b; then ready=false; fi; done
+        if $ready; then break; else echo "..." && sleep 5; fi
+        # Bash does not support boolean operators on true/false and the evaluation of 0 or 1 as true/false changes 
+        # depending on the context and would be much harder to understand
+    done
+
+}
+
+#########################
+# Pre-conditions:
+# - The passed docker container name/ID is valid and running
+# - The command passed is a valid bash command to be executed inside the container
+# - $success_msg is returned by the command upon successful execution
+
+# Post-conditions:
+# - The command is executed successfully with no error code; OR
+# - The command fails $retry_count times in a row and the program exits
+function _docker_exec_with_retry() {
+
+    local retry_count=5
+    local retry_time=(2 4 8 15 30)
+
+    local container=$1
+    local command=$2
+    local success_msg=${3:-""}
+
+    local i=0
+    for ((i; i<retry_count; i++)); do
+        output=$(docker exec "$container" /usr/bin/env bash -c "$command")
+        if [[ $output == *$success_msg* ]]; then
+            return
+        else
+            echo "Command failed, waiting ${retry_time[$i]} seconds before retrying..."
+            sleep "${retry_time[$i]}"
+            
+        fi
+    done
+    echo "Max retries reached while executing $command, $container failed for reason: $output"
+    exit 1
+
+}
+# Pre-conditions: 
+#   - $num_nodes containers running Couchbase Server (uninitialised)/exporter 
+
+# Post-conditions: 
+#   - All CBS/exporter nodes initialised and partitioned as evenly as possible into 
+#     $num_clusters clusters, with a rebalance occurring after the last node is added
+#   - $num_clusters nodes will be running the Data Service, the rest Index/Query, with quotas
+#     specified by $data_alloc and $index_alloc
+#   - Every cluster registered for monitoring with the cbmultimanager
+function configure_servers() {
```

Comment:
I really do think this should all be a custom entrypoint configurable via env vars passed at runtime

Reply:
The reason I say that is then anytime you see a `docker exec...` or `docker cp` command it is generally a code smell and indicative of a non-container approach. I would say we can merge for now but I'd like an improvement JIRA to tidy all this up.

Reply:
Yeah planned for the rewrite, just didn't want to delay things further when porting from the Vagrant example which used `vagrant ssh ...`

Reply:
That's fine, just capture a JIRA.